flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.operator;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.api.misc.param.WithParams;
+import org.apache.flink.ml.params.shared.HasMLEnvironmentId;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableSchema;
+
+import java.io.Serializable;
+
+/**
+ * Base class for algorithm operators.
+ *
+ * <p>Hold a Table as its output.
+ *
+ * <p>For `output` field. The AlgoOperator may have one or more result tables,
+ * in most cases, it has only one result. The output is the main operation result
+ * table, and the other results are kept in the sideOutputs.
+ *
+ * <p>For example:
+ * AlgoA and AlgoB, AlgoB takes the AlgoA’s results as its inputs,
+ * we can write the code like this:
+ * <pre>
+ * {@code
+ * AlgoB.linkFrom(AlgoA)
+ * AlgoA.getOutput()
+ * }
+ * </pre>
+ * The code provides the main result of AlgoA, and AlgoA.getSideOutputs()
+ * provides the other results of AlgoA. AlgoB will take the AlgoA’s results as its
+ * inputs by calling AlgoA.getOutput() and AlgoA.getSideOutputs().
+ *
+ * @param <T> The class type of the {@link AlgoOperator} implementation itself
+ */
+public abstract class AlgoOperator<T extends AlgoOperator <T>>
+	implements WithParams<T>, HasMLEnvironmentId<T>, Serializable {
+
+	/**
+	 * Params for algorithms.
+	 */
+	private Params params;
+
+	/**
+	 * The table held by operator.
+	 */
+	private Table output = null;
+
+	/**
+	 * The side outputs of operator that be similar to the stream's side outputs.
+	 */
+	private Table[] sideOutputs = null;
+
+	/**
+	 * Construct the operator with empty Params.
+	 *
+	 * <p>For how using this constructor, there is an example:
```

Comment:
This constructor is especially useful when users want to set parameters for the algorithm operators. For example, .....

Reply:
Thanks, changed